encoding/migration/convert.go

```diff
@@ -0,0 +1,125 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package migration
+
+import (
+	"fmt"
+
+	"github.com/m3db/m3metrics/encoding"
+	"github.com/m3db/m3metrics/metadata"
+	"github.com/m3db/m3metrics/metric/unaggregated"
+	"github.com/m3db/m3metrics/policy"
+)
+
+func toUnaggregatedMessageUnion(
+	metric unaggregated.MetricUnion,
+	policiesList policy.PoliciesList,
+) (encoding.UnaggregatedMessageUnion, error) {
+	metadatas := toStagedMetadatas(policiesList)
+	switch metric.Type {
+	case unaggregated.CounterType:
+		return encoding.UnaggregatedMessageUnion{
+			Type: encoding.CounterWithMetadatasType,
+			CounterWithMetadatas: unaggregated.CounterWithMetadatas{
+				Counter:         metric.Counter(),
+				StagedMetadatas: metadatas,
+			},
+		}, nil
+	case unaggregated.BatchTimerType:
+		return encoding.UnaggregatedMessageUnion{
+			Type: encoding.BatchTimerWithMetadatasType,
+			BatchTimerWithMetadatas: unaggregated.BatchTimerWithMetadatas{
+				BatchTimer:      metric.BatchTimer(),
+				StagedMetadatas: metadatas,
+			},
+		}, nil
+	case unaggregated.GaugeType:
+		return encoding.UnaggregatedMessageUnion{
+			Type: encoding.GaugeWithMetadatasType,
+			GaugeWithMetadatas: unaggregated.GaugeWithMetadatas{
+				Gauge:           metric.Gauge(),
+				StagedMetadatas: metadatas,
+			},
+		}, nil
+	default:
+		return encoding.UnaggregatedMessageUnion{}, fmt.Errorf("unknown metric type: %v", metric.Type)
+	}
+}
+
+// TODO: look into reuse metadatas during conversion.
+func toStagedMetadatas(
+	policiesList policy.PoliciesList,
+) metadata.StagedMetadatas {
+	numStagedPolicies := len(policiesList)
+	res := make(metadata.StagedMetadatas, 0, numStagedPolicies)
+	for _, sp := range policiesList {
+		if sp.IsDefault() {
+			sm := metadata.DefaultStagedMetadata
+			res = append(res, sm)
+			continue
+		}
+		sm := metadata.StagedMetadata{}
+		sm.CutoverNanos = sp.CutoverNanos
+		sm.Tombstoned = sp.Tombstoned
+		policies, _ := sp.Policies()
+		sm.Metadata = toMetadata(policies)
+		res = append(res, sm)
+	}
+	return res
+}
+
+// TODO: look into reuse metadata during conversion.
+func toMetadata(policies []policy.Policy) metadata.Metadata {
```

Comment:
The overall logic in this function is pretty complex, does the `policies` input have any sort of order? Does the generated `Metadata.Pipelines` need to follow any sort of order? Maybe add some comments for the function?

Reply:
Yeah I can add more comments. The input policies might have been ordered, and if so, the original order is maintained within each pipeline. That said, we actually don't need any ordering on the aggregation tier, so it actually doesn't matter if the output is ordered or not.